cpp/include/cuvs/neighbors/scann.hpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/distance/distance.hpp>
+#include <cuvs/neighbors/common.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/mdspan_types.hpp>
+#include <raft/core/resource/stream_view.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/util/integer_utils.hpp>
+#include <rmm/cuda_stream_view.hpp>
+
+#include <optional>
+#include <variant>
+
+namespace cuvs::neighbors::experimental::scann {
+/**
+ * @defgroup scann_cpp_index_params ScaNN index build parameters
+ * @{
+ */
+
+/**
+ * @brief ANN parameters used by ScaNN to build index
+ *
+ */
+struct index_params : cuvs::neighbors::index_params {
+  // partitioning
+  uint32_t scann_clusters;
+  int64_t kmeans_n_rows_train;
```

Comment:
Is this the absolute number of training rows? Is it what the reference implementation also does? Asking this because in other IVF methods we found that keeping the training rows proportional to number of lists has it's benefits too.

Reply:
[FAISS has 256 as default](https://github.com/facebookresearch/faiss/wiki/How-to-make-Faiss-run-faster#k-means-clustering).  See also [this link](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means). In cuVS we often use 10% of the dataset, but thinking about to change that: https://github.com/rapidsai/cuvs/pull/1125#discussion_r2223959821